### ISSUE TITLE
Favor Pileup over Full_alignment when SNP has high QUAL and Full alignment incorrectly assigns as Refcall

### ISSUE THREAD

**Issue by wshropshire** [HKU-BAL/Clair3#360]
Hello, 

I have a similar issue with #318 where I can see in the VCFs as well as the pileup that under peculiar circumstances, I am seeing near high snp densities where there are pileup variant calls that are correctly assigning an alternate allele but in the full alignment at the same position it's incorrectly typed as as a refcall. Using the same run_clair.sh parameters shown below, I get for a serial isolate with the correct call. This is with bacteria data btw, and I found that in some instances, the GT was being called 0/1 and removed when the evidence for ALT had >.9 AF and high coverage (100X). Here's a picture of the region in particular with the two serial isolates aligned to a reference from a different sequence type and the VCF showing the discordant pileup v. full alignment VCs:

CR-0005 Full Alignment VCF:
<img width="774" alt="Image" src="https://github.com/user-attachments/assets/7bc133b2-2c9d-4139-9516-06e14c89562c" />
CR-0005 Pileup VCF:
<img width="750" alt="Image" src="https://github.com/user-attachments/assets/84ad9c69-5398-4180-8e71-f777cfd45766" />

CR-0063 Full Alignment VCF:
<img width="757" alt="Image" src="https://github.com/user-attachments/assets/a8834ce3-c4d8-472b-a07c-966514d37cb2" />
CR-0063 Pileup VCF:
<img width="773" alt="Image" src="https://github.com/user-attachments/assets/85fa3263-a7bf-479f-93e5-a8b86e9f3047" />

Here's an IGV snapshot:
<img width="677" alt="Image" src="https://github.com/user-attachments/assets/7780d4e7-8c9e-43ed-af42-7f1f4b1d05d5" />

And command here:
`        clair3_cmd = [
            "run_clair3.sh",
            f"--bam_fn={os.path.abspath(bam_output)}",
            f"--ref_fn={os.path.abspath(ref)}",
            f"--threads={threads}",
            "--platform=ont",
            f"--parallel={parallel_path}",
            f"--model_path={os.path.abspath(model_path)}",
            f"--output={clair3_output_dir}",
            f"--sample_name={sample}",
            "--include_all_ctgs",
            "--no_phasing_for_fa",
            "--enable_long_indel"
        ]` 

I can provide data if need be. Love your tool though! This and more consistent masking of recombinant regions and should have a decent core genome alignment pipeline. 

Best, 

Will

**Comment by zhengzhenxian** [comment]
Hi Will, 

Thank you for reporting this. 

It appears that the pileup output provided more reliable results for your bacterial data. The full-alignment model may have failed for some cases due to the high density of SNP variants and high-coverage sample. Given the relatively high-quality scores for the failed cases in pileup, you might consider feeding only a lower proportion of pileup calls into the full-alignment model by using the option `--var_pct_full=0.3` (the default is 0.7).

**Comment by wshropshire** [comment]
Okay, just FYI in your argparse help, you mention the default is 0.3 for `--var_pct_full`. Also, if I increase `--snp_min_af` from 0.08 to 0.9, would that reduce the likelihood of 0/1 genotypes being called and thus allow me to use the `--haploid_precise` option? 

<img width="956" alt="Image" src="https://github.com/user-attachments/assets/4c50ca64-0156-439d-b2f8-1ff2df981cd9" />

**Comment by aquaskyline** [comment]
@wshropshire When the coverage gets higher, it is likely that signals that object a candidate as a true variant get more. Full-alignment was trained to be more sensitive to these signals than the pileup. One might argue that full-alignment should also take the additional supporting signals into account to counterbalance the objecting signals, but a tuning that favors an example might lead to a bunch of failures on other examples. While the input to the Clair3 can be of different species and different coverages, in fact there is no one single perfect solution that fits all usage scenarios. 

Thus we wrote Clair3 to be as modulized as possible. Would you think of writing a small downstream processing script that takes both pileup and full-alignment output as inputs? In the cases you have shown, it seems that you would consider those variants called in pileup, albeit failed the full-alignment, but with good enough VAF and coverage support, as a true variant. You can also give those variants supported by both pileup and full-alignment a higher variant quality, and those supported by only pileup a lower quality, but still consider them as true. What I mean is, a customized downstream processing script gives you the best flexibility, while you can maintain the best recall performance at Clair3's raw outputs.

**Comment by wshropshire** [comment]
Hello Ruibang, 

I'll take your suggestion and try to create a new merge module that takes in a number of conditionals to determine when to favor one call vs another. I frankly don't know how to create de novo QUAL scores so I'm not even going to touch that :). 

How difficult would this be to train the full model with bacterial data? ONT now has bacterial/methylation aware models for dorado basecalling that have significantly improved it's basecalling just in the past year. Could I or clair3 developers think about creating a similar bacteria model?

**Comment by aquaskyline** [comment]
Ya, we kept an eye on any new data for model training. Technically, training a new or refining a model is not too challenging. But the quality of truths need to be really good to lead to a better performance after refinement. Are there any bacteria genomes that are with very good quality truth variants?

**Comment by wshropshire** [comment]
Hey, I have submitted quite a few bacterial genomes to NCBI that I have personally curated myself. I think given that it's been only in the past year I have considered ONT-only assemblies, they all should be hybrid ONT + ILMN assemblies. Regardless, I don't submit assemblies I have worked on unless I've done the proper curation and I can guarantee with some confidence that the assembly is 'error free'. Ryan Wick is another person who would be good to contact about potential 'ground truth' assemblies to train models.

As far as merging the pileup/full alignment models, I've created a merge script that parses through both respective VCFs using pysam and then uses a series of 'ifelse' statements to determine which variant to include in the merged file. It seems to be working fairly well, however, I've kind of hit a dead-end as far as improving results goes. The obstacle now I have narrowed it down to two predominant errors:

 (1) false negative VCs with RefCalls occurring in both models at positions where the alignment indicates good support of the ALT allele (AF > .9) and the AF for the REF Allele is low. Here is an example: 
Incorrect FN in full alignment:
<img width="770" alt="Image" src="https://github.com/user-attachments/assets/edd54a32-f480-45a4-874a-4bd7451f5761" />
Incorrect FN in pileup:
<img width="781" alt="Image" src="https://github.com/user-attachments/assets/806abbbc-8e16-45bd-b1db-5e17fc7c6769" />
Screenshot of IGV pileup:
<img width="241" alt="Image" src="https://github.com/user-attachments/assets/7d5d184f-d06a-46b5-b81f-4be055dd7861" />

Note this occurs near a large deletion

(2) Incorrect false positive VC "PASS" calls in one model, but not the other:
Incorrect call in full alignment:
<img width="775" alt="Image" src="https://github.com/user-attachments/assets/61d702fb-5fff-4ed5-8d7a-6105a153a6ae" />
Correct call in pileup (Despite incorrect AF)
<img width="808" alt="Image" src="https://github.com/user-attachments/assets/8f33deb1-ecdf-4730-95ea-d28c6dd9830a" />
Here's another screenshot of an example near a MNP:
<img width="816" alt="Image" src="https://github.com/user-attachments/assets/db437b8c-31de-4f22-be90-cc01d644bd5c" />

And here's a screenshot where both types of errors are occuring across two different samples:
<img width="1006" alt="Image" src="https://github.com/user-attachments/assets/4d3742f9-f60d-49dd-ba05-e0564314ab12" />

These errors are occurring still predominantly in high SNP density regions and/or INDELs. Since the errors ARE in high density regions, I have been able to mask these regions with Gubbins with partial success, however, that success is still fairly limited.

So I think I've come to the conclusion that the precision/recall of this VF for bacteria really can only improve with better training data specific for bacteria and having a clair3-based model that works with bacteria. I predominantly work with Enterobacterales species, but I also have some other Gram Positive bacteria that I have a graduate student that did both ONT+ILMN assemblies with that you could potentially use. I didn't curate myself, but I can go back and check those as well :)

Let me know if this would be any help. Clair3 has a lot of promise for bacteria and it works well with populations of closely related sequence types, but it still seems to have mostly a specificity issue with more divergent genomes.

**Comment by aquaskyline** [comment]
Agreed with you that having bacteria data included during model training might lead to a better performance. Dropped you an email. I will need a few promising bacteria genomes that have good quality truths and might also contain the high small variant density regions like the cases you've shown.

**Comment by wshropshire** [comment]
Hey Ruibang, I didn't get your email yet, but you can reach me at wcshropshire@mdanderson.org. 

I pinged Ryan Wick on Slack to see if he could help with providing a diverse collection of bacteria from different taxa since mine are almost exclusively belonging to one order of bacteria. Would be still happy to contribute as well :)

**Comment by aquaskyline** [comment]
Just in case, I sent via both my work email '@cs.hku.hk' and gmail.
